specs/guaranteed-gas-market.md

```diff
@@ -0,0 +1,137 @@
+BASE_FEE_MAX_CHANGE_DENOMINATOR = 8
+ELASTICITY_MULTIPLIER = 4
+MAX_RESOURCE_LIMIT = 8,000,000
+TARGET_RESOURCE_LIMIT = MAX_RESOURCE_LIMIT / ELASTICITY_MULTIPLIER
+MINIMUM_BASEFEE=10000
+
+# prev_basefee is a u128, prev_bought_gas and prev_num are u64s
+prev_basefee, prev_bought_gas, prev_num = load_and_unpack_storage()
+now_num = block.number
+
+# Clamp the full basefee to a specific range. The minimum value in the range should be around 100-1000
+# to enable faster responses in the basefee. This replaces the `max` mechanism in the ethereum 1559
+# implementation (it also serves to enable the basefee to increase if it is very small).
+def clamp(v: i256, min: u128, max: u128) -> u128:
+    if v < i256(min):
+        return min
+    elif v > i256(max):
+        return max
+    else:
+        return u128(v)
+
+# If this is a new block, update the basefee and reset the total gas
+# If not, just update the total gas
+if prev_num == now_num:
+    now_basefee = prev_basefee
+    now_bought_gas = prev_bought_gas + requested_gas
+elif prev_num != now_num :
+    # New formula
+    # Width extension and conversion to signed integer math
+    gas_used_delta = int128(prev_bought_gas) - int128(TARGET_RESOURCE_LIMIT)
+    # Use truncating (round to 0) division - solidity's default.
+    # Sign extend gas_used_delta & prev_basefee to 256 bits to avoid overflows here.
+    base_fee_per_gas_delta = prev_basefee * gas_used_delta / TARGET_RESOURCE_LIMIT / BASE_FEE_MAX_CHANGE_DENOMINATOR
+    now_basefee_wide = prev_basefee + base_fee_per_gas_delta
+
+    now_basefee = clamp(now_basefee_wide, min=MINIMUM_BASEFEE, max=UINT_64_MAX_VALUE)
+    now_bought_gas =  requested_gas
+
+# If we skipped multiple blocks between the previous block and now update the basefee again.
+# This is not exactly the same as iterating the above function, but quite close for reasonable
+# gas target values. It is also constant time wrt the number of missed blocks which is important
+# for keeping gas usage stable.
+if prev_num + 1 < now_num:
+    n = now_num - prev_num - 1
+    # Apply 7/8 reduction to prev_basefee for the n empty blocks in a row.
+    now_basefee_wide = prev_basefee * pow(1-(1/BASE_FEE_MAX_CHANGE_DENOMINATOR), n)
+    now_basefee = clamp(now_basefee_wide, min=MINIMUM_BASEFEE, max=UINT_64_MAX_VALUE)
+
+require(now_bought_gas < MAX_RESOURCE_LIMIT)
+
+
+pack_and_store(now_basefee, now_bought_gas, now_num)
+```
+
+### Exponent Based Fee Reduction
+
+When there are stretches where no deposits are executed on L1, the basefee should be decaying, but
+is not. If there is the case that the basefee spiked, this mechanism is needed to enable a more
+accurate decay. It uses exponentiation to run in constant (relative to the number of missed blocks)
+gas.
+
+The current loop based approach will always converge in no more than 12 steps, however it is possible
+to go to a loopless form (with slightly more error), but adpoting a fixed point exponentiation
+algorithm.
```

Comment:
I would just get rid of this whole section and let the algorithm + comments there do the talking — I think it's clear enough.

Also seems outdated since the above uses the exponential decay, and this talks about "the current loop-based approach".

Reply:
Will do. The main loop got updated in the implementation when solmate got released with a new license.